### PR TITLE
[rclcpp] adding a "static" single threaded executor

### PR DESCRIPTION
**Pull request by wjwwood**
This is my rebase of #873, which needed fixup, but I didn't want to push to their fork since they had issues and pull requests on the fork. I can close this and go back to that pull request if possible in the future.

### CODE REVIEW

**Comment by wjwwood** [comment]
@ishugoel20 FYI

You guys can take what's on this branch and force push to your own branch if you like.

**Comment by wjwwood** [comment]
CI:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/9811/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/5438/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/7981/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/9676/)

**Comment by wjwwood** [comment]
Narrower CI on Linux only (just up to `rclcpp`): [Build Status](https://ci.ros2.org/job/ci_linux/9813/)

**Comment by wjwwood** [comment]
I spot checked the CI results with nightly and they all seem unrelated. Also, the rclcpp only job is green. Those circumstances plus the nature of these changes, I'll merge this as-is with good confidence that it will not impact nightlies.

**Comment by ivanpauno** [comment]
There were some clang build warnings that seem to be related with this PR:
- https://ci.ros2.org/job/ci_osx/7981/clang/new/
- https://ci.ros2.org/job/ci_osx/7981/clang-tidy/new/

File: rclcpp/include/rclcpp/strategies/allocator_memory_strategy.hpp [hunk 1]
```diff
@@ -195,6 +194,14 @@ class AllocatorMemoryStrategy : public memory_strategy::MemoryStrategy
     return has_invalid_weak_nodes;
   }
 
+  void add_waitable_handle(const rclcpp::Waitable::SharedPtr & waitable)
```
Comment (emersonknapp) [inline rclcpp/include/rclcpp/strategies/allocator_memory_strategy.hpp]: @wjwwood  this line introduced a build warning on clang https://ci.ros2.org/view/nightly/job/nightly_osx_debug/1577/clang/new/

Comment (mauropasse) [reply rclcpp/include/rclcpp/strategies/allocator_memory_strategy.hpp]: Oh yes, it's missing the `override` keyworkd

Comment (wjwwood) [reply rclcpp/include/rclcpp/strategies/allocator_memory_strategy.hpp]: I'm looking into it.

Comment (wjwwood) [reply rclcpp/include/rclcpp/strategies/allocator_memory_strategy.hpp]: https://github.com/ros2/rclcpp/pull/1040

Comment (wjwwood) [reply rclcpp/include/rclcpp/strategies/allocator_memory_strategy.hpp]: I merged the fix just now, sorry for the disruption.

**Comment by ros-discourse** [comment]
This pull request has been mentioned on **ROS Discourse**. There might be relevant details there:

https://discourse.ros.org/t/improving-ros-2-cpu-by-simplifying-the-ros-2-layers/13808/1

**Comment by ros-discourse** [comment]
This pull request has been mentioned on **ROS Discourse**. There might be relevant details there:

https://discourse.ros.org/t/ros-foxy-fitzroy-released/14495/1

**Pull request merged**